### PR TITLE
Compiler: cleanup compilation of try-catch 

### PR DESCRIPTION
**Pull request by hhugo**
Based on #1342

### CODE REVIEW

File: compiler/lib/generate.ml [hunk 1]
```diff
@@ -674,12 +675,23 @@ let build_graph ctx pc =
...
   let dominance_frontier_cache = Hashtbl.create 17 in
-  let rec loop pc anc =
+  let incr_prec pc =
+    match Hashtbl.find preds pc with
+    | exception Not_found -> Hashtbl.add preds pc 1
+    | n -> Hashtbl.replace preds pc (succ n)
+  in
+  let add_cf_frontier pc front =
+    Addr.Set.iter incr_prec front;
+    cf_frontier := Addr.Map.add pc front !cf_frontier
```
Comment (vouillon) [inline compiler/lib/generate.ml:692]: Could we just add and edge in `succs` from the block containing the `Pushtrap` to right after the `Poptrap` instead?

Comment (hhugo) [reply compiler/lib/generate.ml]: https://github.com/ocsigen/js_of_ocaml/pull/1344/commits/5ba840f8f02f5a263c58d3095f9fb1e727da3d55 does that

**Comment by hhugo** [comment]
rebased on master

**Comment by hhugo** [comment]
@pmwhite, Do you have time to test this PR ?

**Comment by pmwhite** [comment]
@hhugo Yeah, I think I can get to it by next Wednesday, after which I'll be off work for a week.

**Comment by hhugo** [comment]
@vouillon, gentle ping for review

File: compiler/lib/generate.ml [hunk 2]
```diff
@@ -1866,7 +1793,34 @@ and compile_conditional st queue pc last loop_stack backs frontier interm =
...
     | Branch cont -> compile_branch st queue cont loop_stack backs frontier interm
-    | Pushtrap _ -> assert false
+    | Pushtrap (c1, x, e1, _) ->
+        let never_body, body = compile_branch st [] c1 loop_stack backs frontier interm in
+        if debug () then Format.eprintf "@,}@]@,@[<hv 2>catch {@;";
+        let never_handler, handler =
+          compile_branch st [] e1 loop_stack backs frontier interm
+        in
+        let exn_var, handler =
+          assert (not (List.mem x ~set:(snd e1)));
```
Comment (hhugo) [inline compiler/lib/generate.ml:1803]: I'm not sure this assertion always holds.

**Comment by pmwhite** [comment]
Just tested this PR on our codebase. Everything appears to be working!

**Comment by hhugo** [comment]
> Just tested this PR on our codebase. Everything appears to be working!

Thanks for testing

**Comment by hhugo** [comment]
ping @vouillon for more review

**Pull request merged**